### PR TITLE
FALCON-1085 Support Cluster entity updates in Falcon Server

### PR DESCRIPTION
**Pull request by bvellanki**
Added basic documentation, https://issues.apache.org/jira/browse/FALCON-1937 will contain detailed documentation.

### CODE REVIEW

File: webapp/src/test/java/org/apache/falcon/cli/FalconClusterUpdateCLIIT.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        TestContext.deleteEntitiesFromStore();
+    }
+
+
+    public void testUpdateClusterCommands() throws Exception {
+
+        FalconCLI.OUT.set(stream);
+
+        String filePath;
+        TestContext context = new TestContext();
```
Comment (peeyushb) [inline webapp/src/test/java/org/apache/falcon/cli/FalconClusterUpdateCLIIT.java]: "context" has been defined again but it is defined as attribute already.

File: webapp/src/test/java/org/apache/falcon/cli/FalconClusterUpdateCLIIT.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        Map<String, String> overlay = context.getUniqueOverlay();
```
Comment (peeyushb) [inline webapp/src/test/java/org/apache/falcon/cli/FalconClusterUpdateCLIIT.java]: "overlay" is declared as attribute already.

File: prism/src/main/java/org/apache/falcon/resource/AbstractEntityManager.java [hunk 3]
```diff
@@ -359,6 +364,80 @@ protected APIResult update(Entity newEntity, String type, String entityName, Boo
...
+     *
+     * @param clusterName   Name of cluster
+     * @param colo colo
+     * @param skipDryRun Skip dry run during update if set to true
+     * @return APIResult
+     */
+    public APIResult updateClusterDependents(String clusterName, String colo, Boolean skipDryRun) {
+        checkColo(colo);
+        try {
+            verifySafemodeOperation(null, EntityUtil.ENTITY_OPERATION.UPDATE_CLUSTER_DEPENDENTS);
```
Comment (vrangan) [inline prism/src/main/java/org/apache/falcon/resource/AbstractEntityManager.java]: Shouldn't the entityType be a valid entity type?

Comment (bvellanki) [inline prism/src/main/java/org/apache/falcon/resource/AbstractEntityManager.java]: The operation is being performed on either Feed or Process entities. But the actual entity type should not matter for the operation, Hence sending null.

**Comment by vrangan** [comment]
I am giving +1 with the following comments:   Let us fix the minor verifySafemode argument fix so that we are future proof
Let us add JIRA to add versions to other entities.   We initially wanted to have internal managed versions but having exposed them in cluster entity, we should do for others.

File: common/src/main/java/org/apache/falcon/entity/ClusterHelper.java [hunk 4]
```diff
@@ -211,4 +219,55 @@ public static String getEmptyDir(Cluster cluster) {
...
+            for (Property prop : oldEntity.getProperties().getProperties()) {
+                oldProps.put(prop.getName(), prop.getValue());
+            }
+        }
+        if (newEntity.getProperties() != null) {
+            for (Property prop : newEntity.getProperties().getProperties()) {
+                newProps.put(prop.getName(), prop.getValue());
+            }
+        }
+
```
Comment (yzheng-hortonworks) [inline common/src/main/java/org/apache/falcon/entity/ClusterHelper.java]: Better to add a function to return properties of a cluster, as what you did for location and interface.

File: common/src/main/java/org/apache/falcon/entity/ClusterHelper.java [hunk 5]
```diff
@@ -211,4 +219,55 @@ public static String getEmptyDir(Cluster cluster) {
...
+        if (oldProps.keySet().equals(newProps.keySet())) {
+            for (Map.Entry<String, String> keyVals : oldProps.entrySet()) {
+                if (!keyVals.getValue().equals(newProps.get(keyVals.getKey()))) {
+                    return false;
+                }
+            }
+            return true;
```
Comment (yzheng-hortonworks) [inline common/src/main/java/org/apache/falcon/entity/ClusterHelper.java]: Looks you are trying to compare if two maps are identical. You can use the existing function Map::equals (https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#equals(java.lang.Object)).

**Comment by bvellanki** [comment]
@vrangan @peeyushb @yzheng-hortonworks  : Latest patch includes all feedback you provided.

**Comment by peeyushb** [comment]
Patch looks good. +1 .

**Comment by yzheng-hortonworks** [comment]
+1

**Pull request closed without merging**